packages/core/src/fetchers/fetchAllEntities.ts

```diff
@@ -1,66 +1,125 @@
-import { ContentfulClientApi, Entry } from 'contentful';
+import { Asset, ContentfulClientApi, Entry } from 'contentful';
 import { MinimalEntryCollection } from './gatherAutoFetchedReferentsFromIncludes';
 
 const MIN_FETCH_LIMIT = 1;
-
-const fetchEntities = async ({
-  entityType,
+export const fetchAllEntries = async ({
   client,
-  query,
+  ids,
+  locale,
+  skip = 0,
+  limit = 100,
+  responseItems = [],
+  responseIncludes = { Entry: [], Asset: [] },
 }: {
-  entityType: 'Entry' | 'Asset';
   client: ContentfulClientApi<undefined>;
-  query: { 'sys.id[in]': string[]; locale: string; limit: number; skip: number };
+  ids: string[];
+  locale: string;
+  skip?: number;
+  limit?: number;
+  responseItems?: Entry[];
+  responseIncludes?: MinimalEntryCollection['includes'];
 }) => {
-  if (entityType === 'Asset') {
-    return client.getAssets({ ...query });
-  }
+  try {
+    if (!client) {
+      throw new Error(
+        'Failed to fetch experience entities. Required "client" parameter was not provided',
+      );
+    }
+
+    if (!ids.length) {
+      return {
+        items: [],
+        includes: {
+          Entry: [],
+          Asset: [],
+        },
+      };
+    }
+
+    const query = { 'sys.id[in]': ids, locale, limit, skip };
+
+    const {
+      items,
+      includes,
+      total: responseTotal,
+    } = await client.withoutLinkResolution.getEntries({ ...query });
+
+    responseItems.push(...(items as Entry[]));
+    responseIncludes?.Entry?.push(...(includes?.Entry || []));
+    responseIncludes?.Asset?.push(...(includes?.Asset || []));
+
+    if (skip + limit < responseTotal) {
+      await fetchAllEntries({
+        client,
+        ids,
+        locale,
+        skip: skip + limit,
+        limit,
+        responseItems,
+        responseIncludes,
+      });
+    }
 
-  return client.withoutLinkResolution.getEntries({ ...query });
+    return {
+      items: responseItems,
+      includes: responseIncludes,
+    };
+  } catch (error) {
+    if (
+      error instanceof Error &&
+      error.message.includes('size too big') &&
+      limit > MIN_FETCH_LIMIT
+    ) {
+      const newLimit = Math.max(MIN_FETCH_LIMIT, Math.floor(limit / 2));
+      return fetchAllEntries({
+        client,
+        ids,
+        locale,
+        skip,
+        limit: newLimit,
+        responseItems,
+      });
+    }
+
+    throw error;
```

Comment:
When I am reading this, this recursion can happen at the Level0 and at any other levels L1,L2 etc...
This implies that `size too big` message can happen on fetching page3 but not on page1 or page2. 
Also this recursion implies that the halving of the limit can happen recusively eg. `1000->500->250->125->62` (which I guess was the intent. And it would be great to have a test for such a recursive halving too).

But altogether my concern here is that this recursion is two things:
 - monitoring the limit problem 
 - gathering all entries/assets/items

Reply:
I added tests for the recursive halving 

E.g https://github.com/contentful/experience-builder/pull/519/files#diff-8d5e914f2738a65c8156cb1a7020cb49fb345560da7f63c7a7a8773ea7f93aedR144

https://github.com/contentful/experience-builder/pull/519/files#diff-8d5e914f2738a65c8156cb1a7020cb49fb345560da7f63c7a7a8773ea7f93aedR41

Reply:
Also, I don't understand your concern.

If the function is in the process of gathering and it gets the response size too big, it halves the limit else it continues gathering the entities.